### PR TITLE
Create CRD status subresource if it is not present

### PR DESCRIPTION
**Pull request by rossf7**
Towards https://github.com/giantswarm/giantswarm/issues/3293

Extends EnsureCreated in the CRD client to update the CRD if it has a status subresource that is not created. This happens when the CRD is already created for the cluster.

### CODE REVIEW

File: client/k8scrdclient/k8scrdclient.go [hunk 1]
```diff
@@ -80,6 +80,11 @@ func (c *CRDClient) EnsureCreated(ctx context.Context, customResource *apiextens
 		return microerror.Mask(err)
 	}
 
+	err = c.ensureStatusSubresourceCreated(ctx, customResource, backOff)
+	if err != nil {
+		return microerror.Mask(err)
+	}
```
Comment (rossf7) [inline client/k8scrdclient/k8scrdclient.go]: After ensuring the CRD is created we ensure the status subresource is created.

File: client/k8scrdclient/k8scrdclient.go [hunk 2]
```diff
@@ -103,3 +108,37 @@ func (c *CRDClient) EnsureDeleted(ctx context.Context, customResource *apiextens
...
 	return nil
 }
+
+// ensureStatusSubresourceCreated ensures if the CRD has a status subresource
+// it is created. This is needed if a previous version of the CRD without the
+// status subresource is present.
+func (c *CRDClient) ensureStatusSubresourceCreated(ctx context.Context, customResource *apiextensionsv1beta1.CustomResourceDefinition, backOff backoff.BackOff) error {
+	if customResource.Spec.Subresources == nil || customResource.Spec.Subresources.Status == nil {
+		// Nothing to do.
+		return nil
```
Comment (rossf7) [inline client/k8scrdclient/k8scrdclient.go]: If there is no status subresource we can return early.

**Comment by xh3b4sd** [comment]
Pointing to QA. Is this something we want to have e2e tests for? Would that be beneficial?

**Comment by xh3b4sd** [comment]
Makes sense.

**Comment by rossf7** [comment]
> I'll have a go at ~~ending~~ writing a test to giantswarm/e2etests.

LOL. Not sure what I wrote makes sense!

OK I'll proceed here but I'll aim to get the e2e coverage sorted and aligned with our other tests.

File: client/k8scrdclient/k8scrdclient.go [hunk 3]
```diff
@@ -103,3 +108,37 @@ func (c *CRDClient) EnsureDeleted(ctx context.Context, customResource *apiextens
...
+
+	operation := func() error {
+		manifest, err := c.k8sExtClient.ApiextensionsV1beta1().CustomResourceDefinitions().Get(customResource.Name, metav1.GetOptions{})
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		if manifest.Spec.Subresources == nil || manifest.Spec.Subresources.Status == nil {
+			customResource.SetResourceVersion(manifest.ResourceVersion)
+			_, err = c.k8sExtClient.ApiextensionsV1beta1().CustomResourceDefinitions().Update(customResource)
```
Comment (kopiczko) [inline client/k8scrdclient/k8scrdclient.go]: I don't follow. Why do we actually need to call Update here?

Comment (rossf7) [reply client/k8scrdclient/k8scrdclient.go]: We need to call Update if the CRD exists but it doesn't have the status subresource. The current check just creates the CRD if it doesn't exist.

Without this change the chartstatus resource in chart-operator will fail when it calls `UpdateStatus` for the chartconfig CRs.

**Pull request merged**